### ISSUE TITLE
[Question] Working with server side blobs

### ISSUE THREAD

**Issue by kurt-google** [go-goracle/goracle#135]
Ive been looking at using some blobs on the oracle server side during pl/sql but im not sure if goracle supports passing around strictly lob locators. An example of the kind of workflow I hope to achieve:

```go
	db, err := sql.Open("goracle", target)
	if err != nil {
		return err
	}
	defer db.Close()

	tmp := &goracle.Lob{}
	if _, err := db.Exec("DECLARE tmp BLOB; BEGIN dbms_lob.createtemporary(tmp, true, dbms_lob.session); :1 := tmp; END;", goracle.LobAsReader(), sql.Out{Dest: tmp}); err != nil {
		log.Errorf("Failed to create temporary lob: %v", err)
		return err
	}

	buf := bytes.NewBuffer([]byte{1, 2, 3, 4, 5})
	if _, err := db.Exec("BEGIN dbms_lob.append(:1, :2); END;", tmp, goracle.Lob{Reader: buf}); err != nil {
		log.Errorf("Failed to write buffer(%v) to lob(%v): %v", buf.Len(), tmp, err)
		return err
	}

	if _, err := db.Exec("DECLARE BEGIN dbms_lob.freetemporary(:1); END;", tmp); err != nil {
		log.Errorf("Failed to close temporary lob: %v", err)
		return err
	}
	return nil
```

Which fails during the append step with
```
Failed to write buffer(5) to lob(&{0xc0002c5e00 false}): set(data[0][0], goracle.Lob{Reader:(*goracle.dpiLobReader)(0xc0002c5e00), IsClob:false} (goracle.Lob)): getSize: ORA-00000: DPI-1002: invalid OCI handle
```

In this case I know the LOB locator points to a server side lob, and in the 2nd step I just want goracle to pass that locator to be used. However it appears that goracle is trying to read the lob into a temporary and binding that temporary instead of the lob identifier I'm passing in (ref: https://github.com/go-goracle/goracle/blob/master/stmt.go#L1768 ).

Im wondering if Im just misreading this and I might be able to successfully implement this or if this would take changes to goracle to support this kind of server side lob locator support. (I see there are some `Query` related bugs with similar errors caused by the lobs being implicitly closed, perhaps that is happening as well?)

**Comment by tgulacsi** [comment]
The trick is to keep the statement around - db.ExeContext closes the statement right when returned, and this invalidates the LOB handle, too.

So db.PrepareContext + stmt.ExecContext + stmt.Close will work.

**Comment by kurt-google** [comment]
Thanks for responding, just got back to this issue. Even with db.Prepare, stmt.Exec, defer stmt.Close() the 2nd operation still fails but with a new error.

```
Failed to write buffer(5) to lob(&{0xc0003eee40 false}): set(data[0][0], goracle.Lob{Reader:(*goracle.dpiLobReader)(0xc0003eee40), IsClob:false} (goracle.Lob)): closeResource(0x3641ae0): ORA-22289: cannot perform %s operation on an unopened file or LOB
```

This is happening on the `BEGIN dbms_lob.append(:1, :2); END;` statement, im not sure why the %s is in there but seeing as its hitting `closeResource` it seems like its still trying to close a lob (pointer in the error suggests its not my lob?) even when statement closure is deferred?

Reference source
```go
	db, err := sql.Open("goracle", target)
	if err != nil {
		return err
	}
	defer db.Close()

	tmp := &goracle.Lob{}
	stmt, err := db.Prepare("DECLARE tmp BLOB; BEGIN dbms_lob.createtemporary(tmp, true, dbms_lob.session); :1 := tmp; END;")
	if err != nil {
		fmt.Printf("Failed to prepare query: %v\n", err)
		return err
	}
	defer stmt.Close()
	if _, err := stmt.Exec(goracle.LobAsReader(), sql.Out{Dest: tmp}); err != nil {
		fmt.Printf("Failed to create temporary lob: %v\n", err)
		return err
	}

	buf := bytes.NewBuffer([]byte{1, 2, 3, 4, 5})
	stmt, err = db.Prepare("BEGIN dbms_lob.append(:1, :2); END;")
	if err != nil {
		fmt.Printf("Failed to prepare lob append: %v\n", err)
		return err
	}
	defer stmt.Close()
	if _, err := stmt.Exec(tmp, goracle.Lob{Reader: buf}); err != nil {
		fmt.Printf("Failed to write buffer(%v) to lob(%v): %v\n", buf.Len(), tmp, err)
		return err
	}

	stmt, err = db.Prepare("DECLARE BEGIN dbms_lob.freetemporary(:1); END;")
	if err != nil {
		fmt.Printf("Failed to prepare lob free: %v\n", err)
		return err
	}
	defer stmt.Close()
	if _, err := stmt.Exec(tmp); err != nil {
		fmt.Printf("Failed to close temporary lob: %v\n", err)
		return err
	}
	return nil
```

**Comment by tgulacsi** [comment]
For such task, I'd use DBMS_LOB.writeappend.

I don't know for sure, but maybe this error is about the provided `goracle.Lob{Reader:buf}` being closed at the end, but being not a DB-opened LOB?
I'll have to investigate, but till that you could try to skip this `ORA-22289` in that write, and go on (only if you've checked that those bytes have been written for sure).

**Comment by kurt-google** [comment]
I'm not very familiar with oracle's LOB handling honestly, but why would you prefer WRITEAPPEND to APPEND in this case?

**Comment by tgulacsi** [comment]
For append 5 bytes (or any known amounts of bytes), writeappend is better than providing the bytes as a LOB...

**Comment by kurt-google** [comment]
Also thanks for the quick fix, to confirm c797bc7 resolved all the blob handling issues I was having once I started carrying the stmt around with the lob.
